### PR TITLE
core nodes: Template node

### PR DESCRIPTION
**Pull request by ZJvandeWeg**
## Description

This core node use case doc is a bit light compared to others, as the main include docs from upstream are pretty good already.

## Checklist

 - [ ] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [ ] I have considered the performance impact of these changes
 - [ ] Suitable unit/system level tests have been added and they pass
 - [ ] Documentation has been updated

### CODE REVIEW

File: src/_includes/core-nodes/template-use-case.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+You can connect the Template node to a source of data, such as an MQTT input, HTTP input, or any other node that provides data. This input data will be used to populate the template.
+
+### Template Definition
+In the Template node configuration, you define the template using text based formatting languages like HTML with the Mustache syntax.
+Mustache is a simple and "logic-less" templating language that allows you to insert variables and expressions into your template.
+
+For example, you can define a template like this: `<p>Hello {{payload.name}}!</p>`.
+
+![HTML Template with Mustache](./template-mustache.png "HTML template with Mustache")
```
Comment (ZJvandeWeg) [inline src/_includes/core-nodes/template-use-case.md:17]:
```suggestion
![HTML Template with Mustache](./image/template-mustache.png "HTML template with Mustache")
```

File: src/_includes/core-nodes/template-use-case.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+For example, you can define a template like this: `<p>Hello {{payload.name}}!</p>`.
+
+![HTML Template with Mustache](./images/template-mustache.png "HTML template with Mustache")
+
+```json
+[{"id":"97d742eec1cb7dbf","type":"inject","z":"a6b7ede2e13fcbdf","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"name\": \"FlowFuse\"}","payloadType":"json","x":180,"y":60,"wires":[["c85d70b41f374f02"]]},{"id":"c85d70b41f374f02","type":"template","z":"a6b7ede2e13fcbdf","name":"Template using payload.name","field":"payload","fieldType":"msg","format":"handlebars","syntax":"mustache","template":"Hello {{payload.name}}!","output":"str","x":450,"y":60,"wires":[["6df7215459dfb240"]]},{"id":"6df7215459dfb240","type":"debug","z":"a6b7ede2e13fcbdf","name":"Print \"Hello, FlowFuse!\"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","statusVal":"","statusType":"auto","x":730,"y":60,"wires":[]}]
+```
+
+This example will print `Hello FlowFuse!`.
```
Comment (Steve-Mcl) [inline src/_includes/core-nodes/template-use-case.md:23]:
```suggestion
This example will output `<p>Hello FlowFuse!</p>`.
```

**Review by Steve-Mcl — CHANGES_REQUESTED** [review]
It reads fine and is accurate but I feel it could do with more examples and concentrate less on HTML.

For example, it is often used to make small JS objects from a JSON template with `{{mustache}}` values - and set to output parsed JSON. Similarly, it is used to create JS objects from YAML template and `{{mustache}}` values

Finally, sometimes, the template node is used for simply being a placeholder for data (no mustache). Users fill it with CSV or CSS or JSON etc, then use that to deliver content to a HTTP-Response or initiate some processing loop.  We dont mention that mode of operation either.

---


If this is not the focus, then apart from the 1 suggestion, I will approve.

If you are happy for me to do so, I would include a few more examples that folk could use as a reference.

**Comment by Steve-Mcl** [comment]
@ZJvandeWeg 

2 examples I would propose we add:

### The Markdown
```md
    ## More examples

    ### Generating JavaScript objects

    #### input data
    ```js
    {
        payload: ['ACME', 1]
    }
    ```

    #### template node configuration
    * Template:
        ```text
        {
            "product": "{{payload.0}}",
            "version": {{payload.1}}
        }
        ```
    * Format: `Mustache template`
    * Output as: `Parsed JSON`

    #### output
    ```json
    { "product": "ACME", "version": 1 }
    ```

    ### Generating JavaScript object from YAML

    #### input data
    ```js
    {
        topic: 'FlowFuse',
        payload: 3000
    }
    ```

    #### template node configuration
    * Template:
        ```text
        options:
            title: {{topic}}
            port: {{payload}}
        ```
    * Format: `Mustache template`
    * Output as: `Parsed JSON`

    #### output
    ```json
    { "title": "FlowFuse", "port": 3000 }
    ```
```

<br>

### Renders as:
## More examples

### Generating JavaScript objects

#### input data
```js
{
    payload: ['ACME', 1]
}
```

#### template node configuration
* Template:
    ```text
    {
        "product": "{{payload.0}}",
        "version": {{payload.1}}
    }
    ```
* Format: `Mustache template`
* Output as: `Parsed JSON`

#### output
```json
{ "product": "ACME", "version": 1 }
```

### Generating JavaScript object from YAML

#### input data
```js
{
    topic: 'FlowFuse',
    payload: 3000
}
```

#### template node configuration
* Template:
    ```text
    options:
        title: {{topic}}
        port: {{payload}}
    ```
* Format: `Mustache template`
* Output as: `Parsed JSON`

#### output
```json
{ "title": "FlowFuse", "port": 3000 }
```

**Comment by ZJvandeWeg** [comment]
@Steve-Mcl I've added your examples, needed to add the `{%- raw -%}` to make sure Nunjucks wasn't parsing the code examples. Further I've added some comments about comments, and about partials which aren't supported.

File: src/_includes/core-nodes/template-use-case.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+Mustache is a simple and "logic-less" templating language that allows you to insert variables and expressions into your template.
+
+For example, you can define a template like this: `<p>Hello {{payload.name}}!</p>`.
+
+![HTML Template with Mustache](./images/template-mustache.png "HTML template with Mustache")
+
+```json
+[{"id":"97d742eec1cb7dbf","type":"inject","z":"a6b7ede2e13fcbdf","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"name\": \"FlowFuse\"}","payloadType":"json","x":180,"y":60,"wires":[["c85d70b41f374f02"]]},{"id":"c85d70b41f374f02","type":"template","z":"a6b7ede2e13fcbdf","name":"Template using payload.name","field":"payload","fieldType":"msg","format":"handlebars","syntax":"mustache","template":"Hello {{payload.name}}!","output":"str","x":450,"y":60,"wires":[["6df7215459dfb240"]]},{"id":"6df7215459dfb240","type":"debug","z":"a6b7ede2e13fcbdf","name":"Print \"Hello, FlowFuse!\"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","statusVal":"","statusType":"auto","x":730,"y":60,"wires":[]}]
+```
```
Comment (Steve-Mcl) [inline src/_includes/core-nodes/template-use-case.md:22]: doesnt render correctly
image

File: src/_includes/core-nodes/template-use-case.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+Mustache is a simple and "logic-less" templating language that allows you to insert variables and expressions into your template.
+
+For example, you can define a template like this: `<p>Hello {{payload.name}}!</p>`.
+
+![HTML Template with Mustache](./images/template-mustache.png "HTML template with Mustache")
+
+```json
+[{"id":"97d742eec1cb7dbf","type":"inject","z":"a6b7ede2e13fcbdf","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"{\"name\": \"FlowFuse\"}","payloadType":"json","x":180,"y":60,"wires":[["c85d70b41f374f02"]]},{"id":"c85d70b41f374f02","type":"template","z":"a6b7ede2e13fcbdf","name":"Template using payload.name","field":"payload","fieldType":"msg","format":"handlebars","syntax":"mustache","template":"<p>Hello {{payload.name}}!</p>","output":"str","x":450,"y":60,"wires":[["6df7215459dfb240"]]},{"id":"6df7215459dfb240","type":"debug","z":"a6b7ede2e13fcbdf","name":"Print \"<p>Hello, FlowFuse!</p>\"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"payload","targetType":"msg","statusVal":"","statusType":"auto","x":750,"y":60,"wires":[]}]
+```
```
Comment (Steve-Mcl) [inline src/_includes/core-nodes/template-use-case.md:22]: Unfortunately, still NG.

See here: https://deploy-preview-1096--flowforge-website.netlify.app/node-red/core-nodes/template/

I suspect the mustache in line 16 and the JSON flow both need some kind of escape to render the mustache.

image

**Comment by Steve-Mcl** [comment]
@ZJvandeWeg unsure why you have merged this. As it stands, it is inaccurate. see comment  https://github.com/flowforge/website/pull/1096#discussion_r1321289074

**Pull request merged**